api-demo/src/advancedDemo.ts

```diff
@@ -0,0 +1,169 @@
+import * as colors from 'colors';
+import * as os from 'os';
+import * as path from 'path';
+import * as ts from 'typescript';
+import * as tsdoc from '@microsoft/tsdoc';
+
+/**
+ * The advanced demo invokes the TypeScript compiler and extracts the comment from the AST.
+ * It also illustrates how to define custom TSDoc tags using TSDocParserConfiguration.
+ */
+export function advancedDemo(): void {
+  console.log(colors.yellow('*** TSDoc API demo: Advanced Scenario ***') + os.EOL);
+
+  const inputFilename: string = path.resolve(path.join(__dirname, '..', 'assets', 'advanced-input.ts'));
+  const compilerOptions: ts.CompilerOptions = {
+    'target': ts.ScriptTarget.ES5
+  };
+
+  // Compile the input
+  console.log('Invoking the TypeScript compiler to analyze assets/advanced-input.ts...');
+
+  const program: ts.Program = ts.createProgram([ inputFilename ], compilerOptions);
+
+  // Report any compiler errors
+  const compilerDiagnostics: ReadonlyArray<ts.Diagnostic> = program.getSemanticDiagnostics();
+  if (compilerDiagnostics.length > 0) {
+    for (const diagnostic of compilerDiagnostics) {
+      const message: string = ts.flattenDiagnosticMessageText(diagnostic.messageText, os.EOL);
+      if (diagnostic.file) {
+        const location: ts.LineAndCharacter = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!);
+        const formattedMessage: string = `${diagnostic.file.fileName}(${location.line + 1},${location.character + 1}):`
+          + ` [TypeScript] ${message}`;
+        console.log(colors.red(formattedMessage));
+      } else {
+        console.log(colors.red(message));
+      }
+    }
+  } else {
+    console.log('No compiler errors or warnings.');
+  }
+
+  const sourceFile: ts.SourceFile | undefined = program.getSourceFile(inputFilename);
+  if (!sourceFile) {
+    throw new Error('Error retrieving source file');
+  }
+
+  console.log(os.EOL + colors.green('Scanning compiler AST for first code comment...') + os.EOL);
+
+  const foundComments: IFoundComment[] = [];
+
+  walkCompilerAstAndFindComments(sourceFile, '', foundComments);
+
+  if (foundComments.length === 0) {
+    console.log(colors.red('Error: No code comments were found in the input file'));
+  } else {
+    // For the purposes of this demo, only analyze the first comment that we found
+    parseTSDoc(foundComments[0]);
+  }
+}
+
+interface IFoundComment {
+  compilerNode: ts.Node;
+  textRange: tsdoc.TextRange;
+}
+
+function walkCompilerAstAndFindComments(node: ts.Node, indent: string, foundComments: IFoundComment[]): void {
```

Comment:
This function doesn't work in all cases. If a JSDoc-style comment (even just `/** */`) is added before the class declaration, this function fails to find any comments in the source file. I'm not sure what the fix is, but `node.getLeadingTriviaWidth` still seems to work as expected.

Reply:
@RyanCavanaugh Do you know what is the recommended way to traverse a source file and discover `/** */` comments?

Maybe we can improve this demo.

Reply:
@pgonzal You could use the `checker` to get the symbol of a node. Symbols have a method `getDocumentationComment`. Never used this method. The name just sounds fitting.